pkg/runtime/wfengine/workflow.go

```diff
@@ -0,0 +1,251 @@
+	default:
+		wfLogger.Warnf("reminder '%s' for workflow actor '%s' was not recognized", reminderName, actorID)
+	}
+
+	return err
+}
+
+// InvokeTimer implements actors.InternalActor
+func (wf *workflowActor) InvokeTimer(ctx context.Context, actorID string, timerName string, params []byte) error {
+	wfLogger.Debugf("invoking timer '%s' on workflow actor '%s'", timerName, actorID)
+	// TODO
+	return nil
+}
+
+// DeactivateActor implements actors.InternalActor
+func (wf *workflowActor) DeactivateActor(ctx context.Context, actorID string) error {
+	wfLogger.Debugf("deactivating workflow actor '%s'", actorID)
+	delete(wf.states, actorID)
+	return nil
+}
+
+func (wf *workflowActor) createWorkflowInstanceMethod(ctx context.Context, actorID string, startEventBytes []byte) error {
+	// create a new state entry if one doesn't already exist
+	state, exists, err := wf.loadInternalState(actorID)
+	if err != nil {
+		return err
+	} else if !exists {
+		state = &workflowState{
+			inbox: []*backend.HistoryEvent{},
+		}
+	}
+
+	runtimeState := getAndCacheRuntimeState(actorID, state)
+	if exists && !runtimeState.IsCompleted() {
+		return errors.New("workflow instance with this ID already exists")
+	}
+
+	startEvent, err := backend.UnmarshalHistoryEvent(startEventBytes)
+	if err != nil {
+		return err
+	} else if startEvent.GetExecutionStarted() == nil {
+		return errors.New("invalid execution start event")
+	}
+
+	// Schedule a reminder to execute immediately after this operation. The reminder will trigger the actual
+	// workflow execution. This is preferable to using the current thread so that we don't block the client
+	// while the workflow logic is running.
+	if err := wf.createReliableReminder(ctx, actorID, createWorkflowReminder, nil, 0); err != nil {
+		return err
+	}
+
+	state.inbox = []*backend.HistoryEvent{startEvent}
+	return wf.saveInternalState(actorID, state)
+}
+
+func (wf *workflowActor) getWorkflowMetadata(actorID string) (*api.OrchestrationMetadata, error) {
+	state, exists, err := wf.loadInternalState(actorID)
+	if err != nil {
+		return nil, err
+	} else if !exists {
+		return nil, api.ErrInstanceNotFound
+	}
+
+	runtimeState := getAndCacheRuntimeState(actorID, state)
+
+	name, _ := runtimeState.Name()
+	createdAt, _ := runtimeState.CreatedTime()
+	lastUpdated, _ := runtimeState.LastUpdatedTime()
+	input, _ := runtimeState.Input()
+	output, _ := runtimeState.Output()
+	failureDetuils, _ := runtimeState.FailureDetails()
+
+	metadata := api.NewOrchestrationMetadata(
+		runtimeState.InstanceID(),
+		name,
+		runtimeState.RuntimeStatus(),
+		createdAt,
+		lastUpdated,
+		input,
+		output,
+		runtimeState.CustomStatus.GetValue(),
+		failureDetuils,
+	)
+	return metadata, nil
+}
+
+func (wf *workflowActor) runWorkflow(ctx context.Context, actorID string) error {
+	state, exists, err := wf.loadInternalState(actorID)
+	if err != nil {
+		return err
+	} else if !exists {
+		return fmt.Errorf("no workflow state found for actor '%s'", actorID)
```

Comment:
Should we use this return statement on all scenarios where the state does not exist rather than just returning "err"?

Reply:
There is one case where we don't return an error if the state doesn't exist, which is in `createWorkflowInstance` (it's expected to not exist in that case). For that reason, we can't have a single behavior for anytime the state isn't found.